prow/cmd/build/controller_test.go

```diff
@@ -214,6 +235,393 @@ func TestEnqueueKey(t *testing.T) {
 	}
 }
 
+func TestTerminateDupProwJobs(t *testing.T) {
+	now := time.Now()
+	nowFn := func() *metav1.Time {
+		reallyNow := metav1.NewTime(now)
+		return &reallyNow
+	}
+	cases := []struct {
+		name string
+
+		allowCancellations bool
+		pjs                []prowjobv1.ProwJob
+		builds             []buildv1alpha1.Build
+
+		abortedPJs     sets.String
+		expectedBuilds sets.String
+	}{
+		{
+			name:               "terminates all duplicated jobs and all builds",
+			allowCancellations: true,
+			pjs: []prowjobv1.ProwJob{
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "newest", Namespace: fakePJNS},
+					Spec: prowjobv1.ProwJobSpec{
+						Agent: prowjobv1.KnativeBuildAgent,
+						Type:  prowjobv1.PresubmitJob,
+						Job:   "j1",
+						Refs: &prowjobv1.Refs{
+							Repo:  "test",
+							Pulls: []prowjobv1.Pull{{Number: 1}},
+						},
+					},
+					Status: prowjobv1.ProwJobStatus{
+						StartTime: metav1.NewTime(now.Add(-time.Minute)),
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "old", Namespace: fakePJNS},
+					Spec: prowjobv1.ProwJobSpec{
+						Agent: prowjobv1.KnativeBuildAgent,
+						Type:  prowjobv1.PresubmitJob,
+						Job:   "j1",
+						Refs: &prowjobv1.Refs{
+							Repo:  "test",
+							Pulls: []prowjobv1.Pull{{Number: 1}},
+						},
+					},
+					Status: prowjobv1.ProwJobStatus{
+						StartTime: metav1.NewTime(now.Add(-time.Hour)),
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "older", Namespace: fakePJNS},
+					Spec: prowjobv1.ProwJobSpec{
+						Agent: prowjobv1.KnativeBuildAgent,
+						Type:  prowjobv1.PresubmitJob,
+						Job:   "j1",
+						Refs: &prowjobv1.Refs{
+							Repo:  "test",
+							Pulls: []prowjobv1.Pull{{Number: 1}},
+						},
+					},
+					Status: prowjobv1.ProwJobStatus{
+						StartTime: metav1.NewTime(now.Add(-2 * time.Hour)),
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "older-k8s-agent", Namespace: fakePJNS},
+					Spec: prowjobv1.ProwJobSpec{
+						Agent: prowjobv1.KubernetesAgent,
+						Type:  prowjobv1.PresubmitJob,
+						Job:   "j1",
+						Refs: &prowjobv1.Refs{
+							Repo:  "test",
+							Pulls: []prowjobv1.Pull{{Number: 1}},
+						},
+					},
+					Status: prowjobv1.ProwJobStatus{
+						StartTime: metav1.NewTime(now.Add(-2 * time.Hour)),
+					},
+				},
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "completed", Namespace: fakePJNS},
+					Spec: prowjobv1.ProwJobSpec{
+						Agent: prowjobv1.KnativeBuildAgent,
+						Type:  prowjobv1.PresubmitJob,
+						Job:   "j1",
+						Refs: &prowjobv1.Refs{
+							Repo:  "test",
+							Pulls: []prowjobv1.Pull{{Number: 1}},
+						},
+					},
+					Status: prowjobv1.ProwJobStatus{
+						StartTime:      metav1.NewTime(now.Add(-2 * time.Hour)),
+						CompletionTime: nowFn(),
+					},
+				},
+			},
+			builds: []buildv1alpha1.Build{
+				{
+					ObjectMeta: metav1.ObjectMeta{Name: "newest", Namespace: fakePJNS},
```

Comment:
nit:
can we put the braces on  the same line and reduce the vertical space in this file?

Reply:
fixed